{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Postgap GWAS Catalog Input\n",
    "\n",
    "This notebook will generate postgap compatible input summary stats from a GWAS catalog export."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load GWAS Catalog Download\n",
    "\n",
    "From https://www.ebi.ac.uk/gwas/studies/GCST007400 in this case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 71 entries, 0 to 70\n",
      "Data columns (total 38 columns):\n",
      "DATE ADDED TO CATALOG         71 non-null object\n",
      "PUBMEDID                      71 non-null int64\n",
      "FIRST AUTHOR                  71 non-null object\n",
      "DATE                          71 non-null object\n",
      "JOURNAL                       71 non-null object\n",
      "LINK                          71 non-null object\n",
      "STUDY                         71 non-null object\n",
      "DISEASE/TRAIT                 71 non-null object\n",
      "INITIAL SAMPLE SIZE           71 non-null object\n",
      "REPLICATION SAMPLE SIZE       0 non-null float64\n",
      "REGION                        64 non-null object\n",
      "CHR_ID                        64 non-null float64\n",
      "CHR_POS                       64 non-null float64\n",
      "REPORTED GENE(S)              71 non-null object\n",
      "MAPPED_GENE                   64 non-null object\n",
      "UPSTREAM_GENE_ID              19 non-null object\n",
      "DOWNSTREAM_GENE_ID            19 non-null object\n",
      "SNP_GENE_IDS                  45 non-null object\n",
      "UPSTREAM_GENE_DISTANCE        19 non-null float64\n",
      "DOWNSTREAM_GENE_DISTANCE      19 non-null float64\n",
      "STRONGEST SNP-RISK ALLELE     71 non-null object\n",
      "SNPS                          71 non-null object\n",
      "MERGED                        71 non-null int64\n",
      "SNP_ID_CURRENT                65 non-null float64\n",
      "CONTEXT                       64 non-null object\n",
      "INTERGENIC                    71 non-null int64\n",
      "RISK ALLELE FREQUENCY         65 non-null float64\n",
      "P-VALUE                       71 non-null float64\n",
      "PVALUE_MLOG                   71 non-null float64\n",
      "P-VALUE (TEXT)                0 non-null float64\n",
      "OR or BETA                    71 non-null float64\n",
      "95% CI (TEXT)                 71 non-null object\n",
      "PLATFORM [SNPS PASSING QC]    71 non-null object\n",
      "CNV                           71 non-null object\n",
      "MAPPED_TRAIT                  71 non-null object\n",
      "MAPPED_TRAIT_URI              71 non-null object\n",
      "STUDY ACCESSION               71 non-null object\n",
      "GENOTYPING TECHNOLOGY         71 non-null object\n",
      "dtypes: float64(11), int64(3), object(24)\n",
      "memory usage: 21.2+ KB\n"
     ]
    }
   ],
   "source": [
    "# Note that \"OR or BETA\" is OR for this study\n",
    "df = pd.read_csv(\"/Users/eczech/Downloads/gwas-association-downloaded_2020-01-06-accessionId_GCST007400.tsv\", sep=\"\\t\")\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>non_effect_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.0</td>\n",
       "      <td>119503609.0</td>\n",
       "      <td>rs1131265</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.210721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs115268109</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.270027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16.0</td>\n",
       "      <td>68573583.0</td>\n",
       "      <td>rs1170436</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.157004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9.0</td>\n",
       "      <td>99637981.0</td>\n",
       "      <td>rs10819689</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.139262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6072893.0</td>\n",
       "      <td>rs10905718</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.113329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.000000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>8.0</td>\n",
       "      <td>11484361.0</td>\n",
       "      <td>rs2736336</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.292670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.000000e-32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>19.0</td>\n",
       "      <td>1112944.0</td>\n",
       "      <td>rs7249065</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.262364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.000000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>3.0</td>\n",
       "      <td>58332737.0</td>\n",
       "      <td>rs180977001</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.239017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>11.0</td>\n",
       "      <td>128454974.0</td>\n",
       "      <td>rs12575600</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.215111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.000000e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C*05:01</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.235722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.000000e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome     position   variant_id effect_allele non_effect_allele  \\\n",
       "0          3.0  119503609.0    rs1131265                                   \n",
       "1          NaN          NaN  rs115268109                                   \n",
       "2         16.0   68573583.0    rs1170436                                   \n",
       "3          9.0   99637981.0   rs10819689                                   \n",
       "4         10.0    6072893.0   rs10905718                                   \n",
       "..         ...          ...          ...           ...               ...   \n",
       "66         8.0   11484361.0    rs2736336                                   \n",
       "67        19.0    1112944.0    rs7249065                                   \n",
       "68         3.0   58332737.0  rs180977001                                   \n",
       "69        11.0  128454974.0   rs12575600                                   \n",
       "70         NaN          NaN      C*05:01                                   \n",
       "\n",
       "        beta  se       p-value  \n",
       "0   0.210721 NaN  1.000000e-09  \n",
       "1   0.270027 NaN  2.000000e-07  \n",
       "2   0.157004 NaN  9.000000e-07  \n",
       "3   0.139262 NaN  9.000000e-07  \n",
       "4   0.113329 NaN  4.000000e-06  \n",
       "..       ...  ..           ...  \n",
       "66  0.292670 NaN  6.000000e-32  \n",
       "67  0.262364 NaN  9.000000e-06  \n",
       "68  0.239017 NaN  2.000000e-08  \n",
       "69  0.215111 NaN  6.000000e-10  \n",
       "70  0.235722 NaN  9.000000e-09  \n",
       "\n",
       "[71 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The postgap code for loading csvs is poorly written/documented and requires the following fields \n",
    "# for the bayesian finemapping calculations based on position:\n",
    "# chromosome, position, rsID, effect_allele, non_effect_allele, beta, se, pvalue\n",
    "# see: https://github.com/Ensembl/postgap/blob/master/lib/postgap/FinemapIntegration.py#L126\n",
    "# In the main project docs, only the following fields are mentioned as needed per\n",
    "# https://github.com/Ensembl/postgap/blob/master/lib/postgap/GWAS.py#L844:\n",
    "# variant_id, p-value, beta\n",
    "# This transformation will try to satisfy both of those:\n",
    "dfe = df.assign(\n",
    "    effect_allele='', \n",
    "    non_effect_allele='', \n",
    "    se=np.nan,\n",
    "    beta=lambda df: np.log(df['OR or BETA']) # presumably, this is the correct OR -> beta transform\n",
    "    ).rename(columns={\n",
    "        'CHR_ID': 'chromosome', \n",
    "        'CHR_POS': 'position', \n",
    "        'SNPS': 'variant_id',\n",
    "        'OR or BETA': 'odds_ratio',\n",
    "        'P-VALUE': 'p-value'\n",
    "    })[[\n",
    "        'chromosome', 'position', 'variant_id', 'effect_allele', 'non_effect_allele',\n",
    "        'beta', 'se', 'p-value'\n",
    "    ]]\n",
    "dfe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 71 entries, 0 to 70\n",
      "Data columns (total 8 columns):\n",
      "chromosome           64 non-null float64\n",
      "position             64 non-null float64\n",
      "variant_id           71 non-null object\n",
      "effect_allele        71 non-null object\n",
      "non_effect_allele    71 non-null object\n",
      "beta                 71 non-null float64\n",
      "se                   0 non-null float64\n",
      "p-value              71 non-null float64\n",
      "dtypes: float64(5), object(3)\n",
      "memory usage: 4.6+ KB\n"
     ]
    }
   ],
   "source": [
    "dfe.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export --summary_stats arg for postgap\n",
    "dfe.to_csv('/tmp/postgap_GCST007400.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
